Sources/NIOFileSystem/Internal/SystemFileHandle.swift

```diff
@@ -909,6 +923,84 @@ extension SystemFileHandle.SendableView {
 
         return result
     }
+
+    func _setTimes(
+        lastAccess: FileInfo.Timespec?,
+        lastDataModification: FileInfo.Timespec?
+    ) throws {
+        try self._withUnsafeDescriptor { descriptor in
+            let syscallResult: Result<Void, Errno>
+            switch (lastAccess, lastDataModification) {
+            case (.none, .none):
+                // If the timespec array is nil, as per the `futimens` docs,
+                // both the last accessed and last modification times
+                // will be set to now.
+                syscallResult = Syscall.futimens(
+                    fileDescriptor: descriptor,
+                    times: nil
+                )
+
+            case (.some(let lastAccess), .none):
+                // Don't modify the last modification time.
+                syscallResult = Syscall.futimens(
+                    fileDescriptor: descriptor,
+                    times: [timespec(lastAccess), timespec(.omit)]
+                )
+
+            case (.none, .some(let lastDataModification)):
+                // Don't modify the last access time.
+                syscallResult = Syscall.futimens(
+                    fileDescriptor: descriptor,
+                    times: [timespec(.omit), timespec(lastDataModification)]
+                )
+
+            case (.some(let lastAccess), .some(let lastDataModification)):
+                syscallResult = Syscall.futimens(
+                    fileDescriptor: descriptor,
+                    times: [timespec(lastAccess), timespec(lastDataModification)]
+                )
+            }
+
+            try syscallResult.mapError { errno in
+                FileSystemError.futimens(
+                    errno: errno,
+                    path: self.path,
+                    lastAccessTime: lastAccess,
+                    lastDataModificationTime: lastDataModification,
+                    location: .here()
+                )
+            }.get()
+        } onUnavailable: {
+            FileSystemError(
+                code: .closed,
+                message: "Couldn't modify file dates, the file '\(self.path)' is closed.",
+                cause: nil,
+                location: .here()
+            )
+        }
+    }
+}
+
+fileprivate extension timespec {
+    init(_ fileinfoTimespec: FileInfo.Timespec) {
```

Comment:
nit: don't put access modifiers on `extension`s, it's too easy to accidentally give a method the wrong access (this is something we apply to all of our projects)

```suggestion
extension timespec {
    fileprivate init(_ fileinfoTimespec: FileInfo.Timespec) {
```